### ISSUE TITLE
MA0048 false positive on colocated test files: `Foo.Tests.cs` containing `class FooTests`

### ISSUE THREAD

**Issue by Code-Chops** [meziantou/Meziantou.Analyzer#1357]
### Version of the Meziantou.Analyzer NuGet package

3.0.138

### Rule Identifier

MA0048

### Target Framework

net10.0

### C# Language version

C# 14.0

### .NET SDK version

10.0.302

### Description

`MA0048` (File name must match type name) computes the "file name" by truncating at the **first** dot ([`GetFileName` in `FileNameMustMatchTypeNameAnalyzer.cs`](https://github.com/meziantou/Meziantou.Analyzer/blob/main/src/Meziantou.Analyzer/Rules/FileNameMustMatchTypeNameAnalyzer.cs#L145-L158)), then in the default `Exact` mode requires that pre-dot segment to equal the type name.

This produces two behaviors that are inconsistent with each other:

1. Everything after the first dot is ignored for matching, so `Other.AnythingAtAll.cs`containing `class Other` produces **no** diagnostic. Arbitrary dotted suffixes are accepted.
2. The common colocated-test convention, where the dotted suffix is *part of the type name*, is flagged: `Sample.Tests.cs` containing `class SampleTests` warns, as does the Razor variant `Widget.razor.Tests.cs` containing `class WidgetTests`.

So the analyzer is lenient enough to accept any suffix as meaningless, but strict enough to reject the one case where the suffix carries meaning. The `<Subject>.Tests.cs` / `<Subject>Tests` pairing is widespread (colocated tests, `.razor.Tests.cs` for Razor component tests, and similar `.Designer.cs` / `.Generated.cs` style qualifiers), and there is currently no configuration that expresses it short of disabling the rule for those files or switching to `mode = Prefix` (which weakens matching for the suffix-less case too).

## Expected behavior

A type whose name equals the dot-separated file name segments concatenated should be accepted in `Exact` mode, e.g. `Sample` + `Tests` = `SampleTests` for `Sample.Tests.cs`. Ideally lowercase tooling-mandated segments like `razor` may be skipped during concatenation so that `Widget.razor.Tests.cs` accepts `WidgetTests`. An opt-in option (e.g. `MA0048.allow_suffix_concatenation = true`) would also be fine.

## Steps to reproduce

1. Create a class library with `Meziantou.Analyzer` referenced:

   ```xml
   <Project Sdk="Microsoft.NET.Sdk">
     <PropertyGroup>
       <TargetFramework>net10.0</TargetFramework>
     </PropertyGroup>
     <ItemGroup>
       <PackageReference Include="Meziantou.Analyzer" Version="3.0.189">
         <PrivateAssets>all</PrivateAssets>
         <IncludeAssets>runtime; build; native; contentfiles; analyzers</IncludeAssets>
       </PackageReference>
     </ItemGroup>
   </Project>
   ```

2. Add these files (no `.editorconfig`, all defaults):

   `Sample.cs`:
   ```csharp
   namespace Repro;
   public class Sample;
   ```

   `Sample.Tests.cs`:
   ```csharp
   namespace Repro;
   public class SampleTests;
   ```

   `Widget.razor.Tests.cs`:
   ```csharp
   namespace Repro;
   public class WidgetTests;
   ```

   `Other.AnythingAtAll.cs`:
   ```csharp
   namespace Repro;
   public class Other;
   ```

3. `dotnet build`

## Actual output

```
Sample.Tests.cs(3,14): warning MA0048: File name must match type name (class SampleTests), expected file name: 'SampleTests'
Widget.razor.Tests.cs(3,14): warning MA0048: File name must match type name (class WidgetTests), expected file name: 'WidgetTests'
```

`Sample.cs` and, notably, `Other.AnythingAtAll.cs` produce no diagnostic.

## Environment

- Meziantou.Analyzer 3.0.189 (also reproduced on 3.0.138; root cause still present on `main`)
- .NET SDK 10.0.302, macOS 15

## Workarounds

- `[*.Tests.cs]` section with `dotnet_diagnostic.MA0048.severity = none`, losing all validation for test files, or
- `[*.Tests.cs]` with `MA0048.mode = Prefix`, which accepts `SampleTests` in `Sample.Tests.cs` but also accepts it in `Sam.Tests.cs`.

**Comment by meziantou** [comment]
Do you have public examples of projects using `<Subject>.Tests.cs` and not `<Subject>Tests.cs`? Maybe I'm wrong, but it doesn't seem widespread.

**Comment by Code-Chops** [comment]
It's throughout my personal project's solution and all the solution at the company I work in because of colated tests:

<img width="349" height="417" alt="Image" src="https://github.com/user-attachments/assets/a9b4e76c-2095-4833-a860-16b429b34135" />

**Comment by Timovzl** [comment]
I have run into this also, using nesting and segmented names often. I recognize two variants that I think should both be acceptable:
- Using only the first part, as with partial classes: `BlahExtensions.ForStrings.cs` => `class BlahExtensions`
- Using the full name without the dots: `BlahExtensions.ForStrings.Test.cs` => `class BlahExtensionsForStringsTest`

Perhaps the above could be generalized to accepting any prefix of whole segments. I'll leave you to judge that.

**Comment by Code-Chops** [comment]
Thank you very much for adding such a feature so quickly!
